### PR TITLE
Keptn project CNCF proposal

### PR DESCRIPTION
**Pull request by AloisReitbauer**
We want to propose Keptn (www.keptn.sh) as a CNCF sandbox project. Full proposal details will be available soon. We would like to present the project on June 11th.

### CODE REVIEW

**Comment by AloisReitbauer** [comment]
Will next week work for presenting keptn or do you plan for another date? 

Is there also a template for the presentation to use?

**Comment by AloisReitbauer** [comment]
@caniszczyk What do we have to do to move this forward?

**Comment by caniszczyk** [comment]
@AloisReitbauer we will schedule you for a presentation on Aug 13th

https://docs.google.com/presentation/d/1_Fr8DPaKNb-wyM2GQqWC_PtcuatScjnbGFGvySHyAIc/edit#slide=id.g5dce81d046_0_0

You will need to find @cncf/toc members to sponsor before/afterwards to be accepted in the sandbox

**Comment by AloisReitbauer** [comment]
Could we please do the next meeting besides being on my 40th birthday, This is also during my vacation in Mauritius :-)

Thank you for pushing this forward.

**Comment by lizrice** [comment]
It's a little early to say this since the SIG-App Delivery is still forming its charter, but it seems that this would fit within that SIG's remit, and it would be good to get the SIG's view of this project before bringing it to the full TOC

**Comment by AloisReitbauer** [comment]
Flux is also in Sandbox and would fall into the App Delivery charter

File: proposals/keptn.adoc [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+The approach of keptn has a number of advantages over traditional delivery pipelines and operation automation stacks:
+
+* The behavior of a delivery and automation workflow can be easily changed and
+is managed in a central place.
+* Using shipyard and uniform files the "how" is decoupled from the "what" which
+enables greater flexibility in defining operational procedures.
+* The declarative approach allows to even change automation at runtime without
+any need for manual reconfiguration or touching tool integrations.
+
+In the current version keptn focussed on three major use cases for day 1 and day
```
Comment (xiang90) [inline proposals/keptn.adoc]: focussed -> focused

File: proposals/keptn.adoc [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+events are defined as cloud events.
+* The integration layer used knative services to connect the Keptn control plane
+to all required tools to manage application delivery and operations. The
+configuration which tools to use is defined in a keptn uniform file.
+
+The approach of keptn has a number of advantages over traditional delivery pipelines and operation automation stacks:
+
+* The behavior of a delivery and automation workflow can be easily changed and
+is managed in a central place.
+* Using shipyard and uniform files the "how" is decoupled from the "what" which
```
Comment (resouer) [inline proposals/keptn.adoc]: The idea of decoupling strategy from execution makes lots sense to me. 

Though from keptn's documentation, it's not clear how could user control or customize certain strategy. For instance, one user chose strategy of "canary-service" in shipyard.yaml, how could he specify the weight of canary increment step?

**Comment by quinton-hoole** [comment]
Yes, I can confirm that this fits within the scope of the CNCF-App-Delivery SIG.

https://docs.google.com/document/d/1PpHh9D1rE7efR4mX_ClQC1V_piCiP5KMgCpbJp3zMDw/edit

As I understand it, the TOC will finalize the leadership of this SIG in the next week or two, after which the SIG can do a sandbox-level due diligence on this project.

And my apologies, I'd not noticed this PR until now.

**Comment by AloisReitbauer** [comment]
Here you can find the submission slides: https://docs.google.com/presentation/d/1k1QcjeTmb1ZeYtwMntLw1IVTbJXGerGnGExow9oa7i4/edit#slide=id.g60b676c4ff_2_0

File: proposals/keptn.adoc (hunk 1 above)
Comment (AloisReitbauer) [reply proposals/keptn.adoc]: Thx. this should be resolved

File: proposals/keptn.adoc (hunk 2 above)
Comment (dirkwall) [reply proposals/keptn.adoc]: In the `shipyard.yml` you could specify the canary weight increase and the time interval between increases as attributes to that strategy, like

```
strategy: "canary"
- increment: "10%"
- interval: "10m"
```

**Comment by AloisReitbauer** [comment]
Just FYI: I passed the submission to @dirkwall as I am also co-chair of the app delivery sig to avoid any potential conflict of interest.

File: proposals/keptn.adoc [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+for application delivery that can simply be applied across different
+applications.
+
+Keptn solves these problems by providing a configurable control plane to define
+day 1 to day n operations. The design is inspired by SDN stacks:
+
+* The application stack defines how the delivery and operations process should
+work using a shipyard file. Shipyard files define stages, release strategies,
+testing strategies, etc.
+* Keptn than provides a control plane that manages a GitOps repository for the
```
Comment (xiang90) [inline proposals/keptn.adoc]: than -> then

Comment (christian-kreuzberger-dtx) [reply proposals/keptn.adoc]: Fixed.

**Comment by AloisReitbauer** [comment]
This PR has been open since last year May. Should it be converted to the new template. I also assume that given this in "only" a sandbox proposal it should now be moved forward rather quickly.

**Comment by kgamanji** [comment]
Seems like the SIG APP-delivery reviewed the project, but I am not able to find the actual result of the recommendation. Would it be possible to reference the doc?

**Comment by AloisReitbauer** [comment]
Here is the doc: https://docs.google.com/document/d/12Uhkv5aMnvM_HFvpYDuDTN2Zzc83qKOLYmp-k_RGWB8/edit#heading=h.kd4eg2uz3lt0 

Would also be great to have @kgamanji  and @michelleN  as TOC sponsors

**Comment by sheng-liang** [comment]
@AloisReitbauer can you make that Google doc world-readable?

**Comment by amye** [comment]
@AloisReitbauer - GDoc is still locked, thank you!

**Comment by resouer** [comment]
Update: the review comment from SIG App Delivery is added in [Keptn Project Review Doc](https://docs.google.com/document/d/12Uhkv5aMnvM_HFvpYDuDTN2Zzc83qKOLYmp-k_RGWB8/edit#heading=h.6pidoniplc5n).

The general conclusion is SIG App Delivery agrees to recommend Keptn as CNCF sandbox project and it's ready for TOC to sponsor.

Reviewer/Approver from the SIG: Lei Zhang and Bryan Liles.

**Comment by amye** [comment]
Keptn was accepted into the sandbox:
https://lists.cncf.io/g/cncf-toc/message/4853

project onboarding will happen here https://github.com/cncf/toc/issues/474

**Pull request merged**